forge/test/mlir/llama/tests/test_llama_decode.py

```diff
@@ -0,0 +1,321 @@
+    Args:
+        1) use_cache = False
+            input_id (`torch.Tensor`) - shape of (batch_size, seq_length)
+            attention_mask (`torch.Tensor`) -  shape of (batch_size, seq_length)
+        2) use_cache = True
+            input_id (`torch.Tensor`) - shape of (batch_size, 1)
+            attention_mask (`torch.Tensor`) -  shape of (batch_size, key_value_seq_len + 1)
+            position_ids (`torch.Tensor`) -  shape of (batch_size, 1)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`)
+                        key/values shape: (batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)
+
+    Returns:
+        1) use_cache = False
+            logits (`torch.Tensor`) - shape (batch_size, seq_length, vocab_size)
+        2) use_cache = True
+            logits (`torch.Tensor`) - shape (batch_size, 1, vocab_size)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`)
+                        key/values shape: (batch_size, num_of_key_values_heads, key_value_seq_len + 1, head_dim)
+    """
+
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+        self.embed_tokens = model.model.embed_tokens
+
+    def forward(self, input_ids, attention_mask, position_ids=None, past_key_values=None):
+
+        inputs_embeds = self.embed_tokens(input_ids)
+
+        # The causal_mask calculated inside _update_causal_mask method in LlamaModel module
+        # is not properly traced by torch.jit.trace while converting pytorch to relay ops in TVM PyTorch Frontend Conversion
+        # which leads to pcc drops, because the causal mask helps to avoid attention calculation on padding token in LlamaAttention module
+        # To overcome this jit trace issue, causal mask is calculated by using _prepare_4d_causal_attention_mask function.
+        past_key_values_length = past_key_values[0][0].shape[-2] if past_key_values is not None else 0
+        causal_attention_mask = _prepare_4d_causal_attention_mask(
+            attention_mask, input_ids.shape, inputs_embeds, past_key_values_length
+        )
+        model_outputs = self.model(
+            attention_mask=causal_attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+        )
+
+        if model_outputs.past_key_values is not None:
+            return model_outputs.logits, model_outputs.past_key_values
+
+        return model_outputs.logits
+
+
+def calculate_attention_mask_and_postion_ids(
+    padded_past_key_values_seq_length, non_padding_past_key_values_seq_length, input_seq_length
+):
+
+    # Calculate attention mask
+    attention_mask = torch.zeros(padded_past_key_values_seq_length + input_seq_length)
+    attention_mask[:non_padding_past_key_values_seq_length] = 1
+    attention_mask[-1] = 1
+    attention_mask = attention_mask.unsqueeze(0)
+
+    # Calculate position ids
+    position_ids = torch.arange(
```

Comment:
@dgolubovicTT Yes We can calculate position_ids from attention mask but we need to call this prepare_inputs_for_generation method inside the LlamaModelWrapper. After including position_id calculation(i.,e by calling prepare_inputs_for_generation) inside the  LlamaModelWrapper, when I compile the model,  hitting the error in TVM PyTorch frontend conversion. 

```
            # Compile the model on TT
>           compiled_model = forge.compile(framework_model, sample_inputs=[input_ids, attention_mask])

forge/test/mlir/llama/tests/test_llama_decode.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
forge/forge/compile.py:245: in compile_main
    return forge_compile_from_context(compile_context)
forge/forge/compile.py:287: in forge_compile_from_context
    next_stage = stage_to_func[current_stage](context)
forge/forge/compile.py:626: in generate_initial_graph
    module, module_inputs = convert_to_forge_module(
forge/forge/compile.py:980: in convert_to_forge_module
    forge_module, dev_types, module_inputs = generate_forge_module(
forge/forge/tvm_to_python.py:1994: in generate_forge_module
    module_writers, flattened_inputs = compile_tvm_to_python(
forge/forge/tvm_to_python.py:2079: in compile_tvm_to_python
    json_graphs, flattened_pytorch_inputs, weights = load_tvm_graph(
third_party/tvm/python/tvm/contrib/forge_compile.py:106: in load_tvm_graph
    json_graphs, flattened_inputs = compile_tvm_graph(inputs, module, compiler_cfg, graph_name=graph_name, input_names=input_names, path=path, verify_cfg=verify_cfg, framework=framework)
third_party/tvm/python/tvm/contrib/forge_compile.py:146: in compile_tvm_graph
    json_graphs, inputs = compile_pytorch_for_forge(module, *inputs, graph_name=graph_name, compiler_cfg=compiler_cfg, verify_cfg=verify_cfg, input_names=input_names)
third_party/tvm/python/tvm/contrib/forge_compile.py:372: in compile_pytorch_for_forge
    mod, params = tvm.relay.frontend.from_pytorch(traced_model, input_structure, do_convert_params=convert_params)
third_party/tvm/python/tvm/relay/frontend/pytorch.py:6105: in from_pytorch
    outputs = converter.convert_operators(operator_nodes, outputs, ret_name, input_remap)
third_party/tvm/python/tvm/relay/frontend/pytorch.py:5196: in convert_operators
    self.record_output_type(relay_out)
third_party/tvm/python/tvm/relay/frontend/pytorch.py:258: in record_output_type
    self.infer_type_with_prelude(output)
third_party/tvm/python/tvm/relay/frontend/pytorch.py:194: in infer_type_with_prelude
    body = self.infer_type(val, self.prelude.mod)
third_party/tvm/python/tvm/relay/frontend/pytorch.py:187: in infer_type
    new_mod = transform.InferType()(new_mod)
third_party/tvm/python/tvm/ir/transform.py:163: in __call__
    return _ffi_transform_api.RunPass(self, mod)
third_party/tvm/python/tvm/_ffi/_ctypes/packed_func.py:242: in __call__
    raise_last_ffi_error()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def raise_last_ffi_error():
        """Raise the previous error from FFI
    
        This should be used instead of `raise get_last_ffi_error()`, as it
        handle propagation of errors across an FFI boundary.  For example,
        if Python passes a callback to a C++ function, and the callback
        raises an exception, the re-thrown exception should contain the
        full stack trace, not just the stack frames that are above the
        outermost FFI call.
        """
    
        _LIB.TVMGetLastPythonError.restype = ctypes.c_void_p
        _LIB.TVMGetLastBacktrace.restype = ctypes.c_char_p
        py_err = _LIB.TVMGetLastPythonError()
        if py_err is None:
            c_err_msg = py_str(_LIB.TVMGetLastError())
            py_err_msg, err_type = c2pyerror(c_err_msg)
            if err_type is not None and err_type.startswith("tvm.error."):
                err_type = err_type[10:]
            py_err = ERROR_TYPE.get(err_type, TVMError)(py_err_msg)
    
        else:
            # TVMGetLastPythonError returns a PyObject*, with NULL when
            # there is no such value.  If we annotated the restype as
            # ctypes.py_object, we would need to return Py_None from the
            # C++ implementation.  This would require introducing a
            # dependency on libpython that we want to avoid when not in a
            # Python environment.  Therefore, casting the resulting void*
            # pointer to PyObject* using ctypes.
            py_err = ctypes.cast(ctypes.c_void_p(py_err), ctypes.py_object).value
    
        tb = py_err.__traceback__
    
        # The py_err.__traceback__ only goes from the location thrown
        # up to the next FFI handoff.  To have the stacktrace also
        # include the C++ side, we need to adjust the __traceback__
        # before re-throwing.
        backtrace = _LIB.TVMGetLastBacktrace()
        if backtrace:
            frames = re.split(r"\n\W+\d+:\W+", py_str(backtrace))
            frames = frames[1:]  # Skip "Stack trace: "
    
            for frame in frames:
                if " at " in frame:
                    func_name, frame = frame.split(" at ", 1)
                    filename, lineno = frame.rsplit(":", 1)
                    func_name = func_name.strip()
                    filename = filename.strip()
                    lineno = int(lineno.strip())
    
                    tb = _append_traceback_frame(tb, func_name, filename, lineno)
    
        # Remove stack frames that provide little benefit to
        # debugging.  These are only removed from the stack frames
        # contained within the exception we are re-raising, and not to
        # the stack frames that it will continue to collect.
        # Therefore, there may still be a single instance of these
        # frames in the outermost Python-to-FFI call.
        filter_funcs = [
            lambda code: "tvm/_ffi/_ctypes/packed_func.py" in code.co_filename,
            lambda code: "tvm/_ffi/base.py" in code.co_filename,
        ]
        tb = _filter_traceback_frames(tb, filter_funcs)
    
        py_err = py_err.with_traceback(tb)
    
        # The exception PyObject may contain a large amount of state,
        # including all stack frames that may be inspected in a later
        # PDB post-mortem.  Therefore, we must make sure to remove the
        # underlying PyObject* from the C++ side after we retrieve it.
        _LIB.TVMDropLastPythonError()
    
>       raise py_err
E       tvm._ffi.base.TVMError: Traceback (most recent call last):
E         8: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<tvm::IRModule (tvm::transform::Pass, tvm::IRModule)>::AssignTypedLambda<tvm::transform::__mk_TVM9::{lambda(tvm::transform::Pass, tvm::IRModule)#1}>(tvm::transform::__mk_TVM9::{lambda(tvm::transform::Pass, tvm::IRModule)#1}, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tvm::runtime::TVMRetValue)
E         7: tvm::transform::Pass::operator()(tvm::IRModule) const
E         6: tvm::transform::Pass::operator()(tvm::IRModule, tvm::transform::PassContext const&) const
E         5: tvm::transform::ModulePassNode::operator()(tvm::IRModule, tvm::transform::PassContext const&) const
E         4: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<tvm::IRModule (tvm::IRModule, tvm::transform::PassContext)>::AssignTypedLambda<tvm::relay::transform::InferType()::{lambda(tvm::IRModule, tvm::transform::PassContext const&)#1}>(tvm::relay::transform::InferType()::{lambda(tvm::IRModule, tvm::transform::PassContext const&)#1})::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)
E         3: tvm::DiagnosticContext::Render()
E         2: tvm::DiagnosticRenderer::Render(tvm::DiagnosticContext const&)
E         1: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<void (tvm::DiagnosticContext)>::AssignTypedLambda<tvm::TerminalRenderer(std::ostream&)::{lambda(tvm::DiagnosticContext const&)#1}>(tvm::TerminalRenderer(std::ostream&)::{lambda(tvm::DiagnosticContext const&)#1})::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)
E         0: tvm::ReportAt(tvm::DiagnosticContext const&, std::ostream&, tvm::Span const&, tvm::Diagnostic const&)
E         File "/proj_sw/user_dev/pchandrasekaran/latest/Forge/tt-forge-fe/third_party/tvm/src/ir/diagnostic.cc", line 267
E       TVMError: The source maps are not populated for this module. Please use `tvm.relay.transform.AnnotateSpans` to attach source maps for error reporting.
E       Error: The Relay type checker is unable to show the following types match.
E       In particular `Tensor[(1, 58), float32]` does not match `Tensor[(1, 58), int64]`

```
The error is raised because of the position_id calculation inside the LlamaModelWrapper 
Repro:
```
def test_postion_id_calculation():
    
    class PositionIdModelWrapper(torch.nn.Module):
        def __init__(self):
            super().__init__()
            
        def forward(self, attention_mask):
            position_ids = attention_mask.long().cumsum(-1) - 1
            position_ids.masked_fill_(attention_mask == 0, 1)
            return position_ids
     
    framework_model = PositionIdModelWrapper()
    framework_model.eval()
    
    attention_mask = torch.zeros((1, 20)).to(torch.int64)
    attention_mask[: , :12] = 1

    framework_out = framework_model(attention_mask)
    compiled_model = forge.compile(framework_model, sample_inputs=[attention_mask])
```

Reply:
Sure Will create a issue